src/vos/vos_obj_cache.c

```diff
@@ -225,19 +225,17 @@ vos_obj_hold(struct daos_lru_cache *occ, struct vos_container *cont,
 			obj->obj_epoch = epoch;
 
 		if (intent == DAOS_INTENT_KILL) {
-			if (!obj->obj_df) /* new object, nothing to delete */
-				D_GOTO(failed, rc = -DER_NONEXIST);
-
 			if (vos_obj_refcount(obj) > 2)
 				D_GOTO(failed, rc = -DER_BUSY);
 
 			/* no one else can hold it */
 			obj->obj_zombie = true;
 			vos_obj_evict(obj);
-			break; /* OK to delete */
+			if (obj->obj_df)
+				goto out; /* OK to delete */
```

Comment:
we don't necessarily fall through? we can just decref and return NONEXIST if obj->obj_df == NULL?

Reply:
That is the point, I think we do need to fall through so we call vos_oi_find on the object.   obj_df == NULL doesn't mean the object doesn't exist at this point.  It just means it's not in cache at present.

For instance, if the server opens a pool and then calls delete on an existing object.  The object will not be in cache so the call to lru_hold will create the cache entry and obj_df will be NULL here.  We would want it to lookup the object in the persistent memory here before we assume it doesn't exist.